### PR TITLE
Extending Exit Image and Adding Key frames

### PR DESCRIPTION
**Pull request by Oncorporation**
- Make Exit Frame solid like a key frame
- Key frames: allow image files to be specified in the JSON and DataFrame
Example: https://youtu.be/4aUZhXmhd_s

Notes: Unsure if we need exit and init frames now.

Example of a JSON file:
[sweet-dreams-infinite-zoom-prompts12-TEST-inverted json.txt](https://github.com/v8hid/infinite-zoom-automatic1111-webui/files/11302065/sweet-dreams-infinite-zoom-prompts12-TEST-inverted.json.txt)

### CODE REVIEW

**Comment by Oncorporation** [comment]
I understand what I did, so the conflict looks simple to me.

Just use my code because it is adding some new variables in the logically flow. It isn't a merge.

This allows the exit image and the keyframe images to not immediately have an images placed in their center.

**Comment by v8hid** [comment]
I'm refactoring the whole thing, you have conflict here so, we probably have to merge this to the refactored version later.

**Comment by Oncorporation** [comment]
Refactored!

File: iz_helpers/static_variables.py
Comment (GeorgLegato) [inline iz_helpers/static_variables.py:10]: Instead of a filesystem path, lets consider a DATAURL. Browser and fileystem are not good friends.
and, when sharing promps none has the images being referenced by this path.
the dataurl is base64 encoded and keeps the image embedded. sure, the json grows. but thats ok

File: iz_helpers/static_variables.py
Comment (GeorgLegato) [inline iz_helpers/static_variables.py:21]: if using dataurl we could paste an example image "data:png;base64: codebla.."

**Review by GeorgLegato — APPROVED** [review]

File: iz_helpers/promptschema.json
Comment (Oncorporation) [reply iz_helpers/promptschema.json:42]: I was providing blank strings to ensure everything imports, maybe you are right and they should not be required.  I will test.

File: iz_helpers/static_variables.py
Comment (Oncorporation) [reply iz_helpers/static_variables.py:10]: That would be a huge JSON file. Maybe we could make it compatible with both as it is a string. We can be confident that we are in a local installed environment, so the filesystem path works in my projects.

I am sure we can make or find a function that reads, filesystem, URL and DataURL.

Comment (Oncorporation) [reply iz_helpers/static_variables.py:10]: Done

Comment (Oncorporation) [reply iz_helpers/static_variables.py:10]: [infinite-zoom-prompts-12-test-with-images.json.txt](https://github.com/v8hid/infinite-zoom-automatic1111-webui/files/11307527/infinite-zoom-prompts-12-test-with-images.json.txt)
Successfully Tested this with :
 a file path like "F:\\sd\\stable-diffusion-webui\\outputs\\txt2img-images\\myimage.png"
-  a URL like "http:www.openai.com/images/myimage.png" 
- a DataURL like "data:png;base64: codebla.."

Comment (v8hid) [reply iz_helpers/static_variables.py:10]: > We can be confident that we are in a local installed environment

as the maintainer of the repo I'm using a server because I don't have a local GPU like so many other people, so we should not consider everyone on local env

**Comment by v8hid** [comment]
Hi @Oncorporation and @GeorgLegato,

I hope you're doing well. I noticed that there may be some features being developed that might not be necessary at this time.

I would like to suggest that we focus on keeping things simple for now and avoid implementing anything beyond the initial image and exit image. This will allow us to prioritize other improvements that may be more critical to the project's success.

I believe it would be beneficial for us to discuss this further as a team and explore different ideas and approaches. I'm open to hearing your thoughts and suggestions as well.

Thank you for your hard work and dedication to this project. I'm looking forward to working together to make it a success.

**Comment by GeorgLegato** [comment]
> Hi @Oncorporation and @GeorgLegato,
> 
> I hope you're doing well. I noticed that there may be some features being developed that might not be necessary at this time.
> 
> I would like to suggest that we focus on keeping things simple for now and avoid implementing anything beyond the initial image and exit image. This will allow us to prioritize other improvements that may be more critical to the project's success.
> 
> I believe it would be beneficial for us to discuss this further as a team and explore different ideas and approaches. I'm open to hearing your thoughts and suggestions as well.
> 
> Thank you for your hard work and dedication to this project. I'm looking forward to working together to make it a success.

you right, @v8hid , esp. the gradio table is currently bad for handling additional images. pasting text like URL or data is not useable. @Oncorporation let us keep this PR on hold until we get a "better table"; people asking for other UI anyways.
If we can add then an img-element (dragdrop, preview, delete etc) then let us merge this into.
I see the exit image only and a fix on smoothness in deed more prior.

File: iz_helpers/static_variables.py
Comment (Oncorporation) [reply iz_helpers/static_variables.py:10]: Could you use a 4090 or do you prefer to work remotely?

I am unsure how you accomplished being on a server at a reasonable price.

**Comment by Oncorporation** [comment]
> 

Agreed. 

However, for my goals I need a JSON file that can be imported with the images or prompts included.  The string based image handling is perfect for my use case because including entire images would be make the JSON files cumbersome and the image storage would be redundant.

Even in a colab, relative image locations (google drive) are likely to work with minimal effort. 

It may be possible to put a component in the dataframe for choosing a file. However, I have not attempted it. 

The current limitations of the dataframe input may be solved with css or other formatting.

**Comment by v8hid** [comment]
I've Changed the base branch to exit_image

**Pull request merged**
